### PR TITLE
POC: initial reflection approach commit

### PR DESCRIPTION
**Pull request by stijnmoreels**
Reflection approach: extracting the open message handling interface from the `IServiceProvider` and dynamically call the `.ProcessMessageAsync` method on the first message handler that conforms to the message contract of the incoming raw message body.

### CODE REVIEW

**Comment by stijnmoreels** [comment]
Of course not ready to merge on the remote arcus.messaing branch, but ready to be viewed as a possible approach.
@tomkerkhove

File: src/Arcus.Messaging.Tests.Workers/MessageHandlers/OrdersMessageHandler.cs [hunk 1]
```diff
@@ -34,33 +27,25 @@ public OrdersMessageHandler(IEventGridPublisher eventGridPublisher, ILogger<Orde
...
         /// <param name="messageContext">Context providing more information concerning the processing</param>
         /// <param name="correlationInfo">
         ///     Information concerning correlation of telemetry & processes by using a variety of unique
         ///     identifiers
         /// </param>
         /// <param name="cancellationToken">Cancellation token</param>
-        public async Task<MessageProcessResult> ProcessMessageAsync(
-            Message message, 
+        public async Task ProcessMessageAsync(
+            Order order,
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Tests.Workers/MessageHandlers/OrdersMessageHandler.cs:38]: 👌

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 2]
```diff
@@ -283,7 +284,10 @@ private async Task HandleMessageAsync(Message message, CancellationToken cancell
                 var messageContext = new AzureServiceBusMessageContext(message.MessageId, message.SystemProperties,
                     message.UserProperties);
 
-                await ProcessMessageAsync<IAzureServiceBusMessageHandler>(message, messageContext, correlationInfo, cancellationToken);
+                Encoding encoding = DetermineMessageEncoding(messageContext);
+                string messageBody = encoding.GetString(message.Body);
+
+                await ProcessMessageAsync(messageBody, messageContext, correlationInfo, cancellationToken);
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs:290]: 👌

File: src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs [hunk 3]
```diff
@@ -81,81 +81,71 @@ protected virtual Task HandleReceiveExceptionAsync(Exception receiveException)
...
         }
 
+        // TODO: move to extension?
+        /// <summary>
+        ///     Determines the encoding used for a given message
+        /// </summary>
+        /// <remarks>If no encoding was specified, UTF-8 will be used by default</remarks>
+        /// <param name="messageContext">Context concerning the message</param>
+        /// <returns>Encoding that was used for the message body</returns>
+        protected virtual Encoding DetermineMessageEncoding(MessageContext messageContext)
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs:91]: Don't we already have this?

AFAIK I already wrote this

File: src/Arcus.Messaging.Pumps.Abstractions/MessageHandling/MessageHandler.cs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+using GuardNet;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Arcus.Messaging.Pumps.Abstractions.MessageHandling
+{
+    /// <summary>
+    /// Represents an abstracted form of the <see cref="IMessageHandler{TMessage,TMessageContext}"/> implementation to handle with different type of generic message and message context types.
+    /// </summary>
+    internal class MessageHandler
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.Abstractions/MessageHandling/MessageHandler.cs:18]: All the rest looks fine but this is a lot of voodoo, let's discuss in our meeting next week

File: src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs (hunk 3 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs:91]: Yes, was only removed in the destination PR-branch. Now reverted. I added a TODO though to maybe move this out of the message pump.

**Pull request merged**